KsApi/combine/CombineTestObserver.swift

```diff
@@ -27,4 +33,161 @@ public final class CombineTestObserver<Value, Error: Swift.Error> {
+      }
+    }
+
+    return values
+  }
+
+  /// Get the last value emitted by the signal.
+  public var lastValue: Value? {
+    return self.values.last
+  }
+
+  /// `true` if at least one `.Next` value has been emitted.
+  public var didEmitValue: Bool {
+    return self.values.count > 0
+  }
+
+  /// The failed error if the signal has failed.
+  public var failedError: Error? {
+    var errors: [Error] = []
+    for event in self.events {
+      switch event {
+      case let .error(e):
+        errors.append(e)
+      default: break
+        // do nothing
+      }
+    }
+
+    assert(
+      errors.count <= 1,
+      "I'm pretty sure a Combine publisher can only ever emit one error. If this fails, we've learned something new today."
+    )
+
+    return errors.last
+  }
+
+  /// `true` if a `.Failed` event has been emitted.
+  public var didFail: Bool {
+    return self.failedError != nil
+  }
+
+  /// `true` if a `.Finished` event has been emitted or a `.Failed` event has been ommitted
+  public var didComplete: Bool {
+    return self.events.contains { event in
+      switch event {
+      case .finished:
+        return true
+      case .error:
+        return true
+
+      default: break
+      }
+      return false
+    }
+  }
+
+  public func assertDidComplete(_ message: String = "Should have completed.",
+                                file: StaticString = #file, line: UInt = #line) {
+    XCTAssertTrue(self.didComplete, message, file: file, line: line)
+  }
+
+  public func assertDidFail(_ message: String = "Should have failed.",
+                            file: StaticString = #file, line: UInt = #line) {
+    XCTAssertTrue(self.didFail, message, file: file, line: line)
+  }
+
+  public func assertDidNotFail(_ message: String = "Should not have failed.",
+                               file: StaticString = #file, line: UInt = #line) {
+    XCTAssertFalse(self.didFail, message, file: file, line: line)
+  }
+
+  public func assertDidNotComplete(_ message: String = "Should not have completed",
+                                   file: StaticString = #file, line: UInt = #line) {
+    XCTAssertFalse(self.didComplete, message, file: file, line: line)
+  }
+
+  public func assertDidEmitValue(_ message: String = "Should have emitted at least one value.",
+                                 file: StaticString = #file, line: UInt = #line) {
+    XCTAssert(self.values.count > 0, message, file: file, line: line)
+  }
+
+  public func assertDidNotEmitValue(_ message: String = "Should not have emitted any values.",
+                                    file: StaticString = #file, line: UInt = #line) {
+    XCTAssertEqual(0, self.values.count, message, file: file, line: line)
+  }
+
+  public func assertDidTerminate(
+    _ message: String = "Should have terminated, i.e. completed/failed/interrupted.",
+    file: StaticString = #file, line: UInt = #line
+  ) {
+    XCTAssertTrue(self.didFail || self.didComplete, message, file: file, line: line)
```

Comment:
Nit: I don't think you need `self.didFail` here - if it failed `self.didComplete` will return true as well. Same for `didNotTerminate` below

Reply:
Legit! This is copy-pasted from the original `TestObserver` so I'm inclined to just leave it for parity.